src/SA/listmods/entry.c

```diff
@@ -0,0 +1,130 @@
+#include <windows.h>
+#include <stdio.h>
+#include "psapi.h"
+#include "bofdefs.h"
+#include "base.c"
+
+DECLSPEC_IMPORT LPVOID WINAPI KERNEL32$OpenProcess(DWORD dwDesiredAccess, BOOL bInheritHandle, DWORD dwProcessId);
+DECLSPEC_IMPORT DWORD KERNEL32$GetCurrentProcessId();
+DECLSPEC_IMPORT HGLOBAL KERNEL32$GlobalAlloc(UINT uFlags, SIZE_T dwBytes);
+DECLSPEC_IMPORT HGLOBAL KERNEL32$GlobalFree(HGLOBAL hMem);
+DECLSPEC_IMPORT WINBOOL WINAPI PSAPI$EnumProcessModulesEx(HANDLE hProcess, HMODULE *lphModule, DWORD cb, LPDWORD lpcbNeeded, DWORD dwFilterFlag);
+DECLSPEC_IMPORT DWORD PSAPI$GetModuleFileNameExA(HANDLE hProcess, HMODULE hModule, LPSTR lpFilename, DWORD nSize);
+
+typedef DWORD GetFileVersionInfoSizeA_t(
+  LPCSTR  lptstrFilename,
+  LPDWORD lpdwHandle
+);
+
+typedef WINBOOL GetFileVersionInfoA_t(
+  LPCSTR lptstrFilename,
+  DWORD  dwHandle,
+  DWORD  dwLen,
+  LPVOID lpData
+);
+
+typedef WINBOOL VerQueryValueA_t(
+  LPCVOID pBlock,
+  LPCSTR  lpSubBlock,
+  LPVOID  *lplpBuffer,
+  PUINT   puLen
+);
+
+// Inspired from: https://www.codeguru.com/cpp/w-p/win32/versioning/article.php/c4539/Versioning-in-Windows.htm
+int PrintSingleModule(char* szFile){
+    DWORD dwLen, dwUseless;
+    LPTSTR lpVI;
+    char* companyName;
+
+    GetFileVersionInfoSizeA_t* GetFileVersionInfoSizeA = (GetFileVersionInfoSizeA_t*) GetProcAddress(LoadLibraryA("Api-ms-win-core-version-l1-1-0.dll"), "GetFileVersionInfoSizeA");
+    dwLen = GetFileVersionInfoSizeA((LPTSTR)szFile, &dwUseless);
+    if (dwLen==0){
+        internal_printf("%-50sERROR: Could not GetFileVersionInfoSizeA() on the DLL.\n", szFile);
+        return 1;
+    }
+
+    lpVI = (LPTSTR) KERNEL32$GlobalAlloc(GPTR, dwLen);
+    if (lpVI) {
+        DWORD dwBufSize;
+        VS_FIXEDFILEINFO* lpFFI;
+        BOOL bRet = FALSE;
+        WORD* langInfo;
+        UINT cbLang;
+        char szVerDescription[128];
+        char szVerCompanyName[128];
+        LPVOID lpDescription;
+        LPVOID lpCompanyName;
+        UINT cbBufSize;
+
+        GetFileVersionInfoA_t* GetFileVersionInfoA = (GetFileVersionInfoA_t*) GetProcAddress(LoadLibraryA("Api-ms-win-core-version-l1-1-0.dll"), "GetFileVersionInfoA");
+        GetFileVersionInfoA((LPTSTR)szFile, 0, dwLen, lpVI);
+
+        //First, to get string information, we need to get
+        //language information.
+        VerQueryValueA_t* VerQueryValueA = (VerQueryValueA_t*) GetProcAddress(LoadLibraryA("Api-ms-win-core-version-l1-1-0.dll"), "VerQueryValueA");
+        VerQueryValueA(lpVI, "\\VarFileInfo\\Translation", (LPVOID*)&langInfo, &cbLang);
+
+        // Get Description
+        MSVCRT$sprintf(szVerDescription, "\\StringFileInfo\\%04x%04x\\%s", langInfo[0], langInfo[1], "FileDescription");
```

Comment:
these can be converted to just internal_printf, so you don't have to stage through a secondary variable

Reply:
Not sure I can do this. The first `VerQueryValueA` will query for lang information. The lang is later used to build other strings that act as query strings. They won't be printed.

Reply:
Ah I see that now, sorry I missed that in my first pass